shell/common/pipeline.h

```diff
@@ -145,18 +179,33 @@ class Pipeline : public fml::RefCountedThreadSafe<Pipeline<R>> {
   fml::Semaphore available_;
   std::mutex queue_mutex_;
   std::queue<std::pair<ResourcePtr, size_t>> queue_;
+  PipelineStateObserver pipeline_state_observer_;
 
   void ProducerCommit(ResourcePtr resource, size_t trace_id) {
+    bool had_resource = !!resource;
+
     {
       std::lock_guard<std::mutex> lock(queue_mutex_);
       queue_.emplace(std::move(resource), trace_id);
     }
 
     // Ensure the queue mutex is not held as that would be a pessimization.
     available_.Signal();
+
+    if (pipeline_state_observer_) {
```

Comment:
I think the resource is always valid. You can probably just assert that here. I case of an aborted item, I believe that item is just dropped on the floor. So I don't actually think you will ever get an abort.

Reply:
If you add these state observer fire events right next to the calls to TRACE_EVENT_ASYNC BEGIN, STEP and END, you should get the right behavior.